### PR TITLE
feat: add column_type and influxdb_column_type, remove row_count from system.columns

### PR DESCRIPTION
**Pull request by alamb**
Closes https://github.com/influxdata/influxdb_iox/issues/1413

# Rationale:

I want to be able to programatically interrogate the schema of tables in IOx via SQL to know if they are tags or fields (or timestamps, but that is fairly obvious).

Also, the intent is that `system.columns` contains IOX schema information, not info about the data in those tables.

Note:
1. row_counts are already exposed in `system.chunks`, and I plan to expose the full information in ColumnSummary in for https://github.com/influxdata/influxdb_iox/issues/1340
2. Arrow / SQL types are already exposed in `information_schema.columns` (or `SHOW COLUMNS`) so keep `system.columns` showing IOx specific information

# Changes

1. Add `column_type` and `influxdb_column_type` to `system.columns`
2. Remove `row_count` column from `system.columns`
2. Add  additional explination of the system tables in the documentation

### CODE REVIEW

File: data_types/src/partition_metadata.rs [hunk 1]
```diff
@@ -107,10 +111,34 @@ impl TableSummary {
     }
 }
 
+// Replicate this enum here as it can't be derived from the existing statistics
+#[derive(Debug, Deserialize, Serialize, PartialEq, Clone)]
```
Comment (alamb) [inline data_types/src/partition_metadata.rs:115]: I am not thrilled about adding yet another enum for Tag/Field/Timestamp (as there is already one in `internal_types` and one (maybe more) in `read_buffer`) but given that the `data_types` crate doesn't depend on any of those others, this seems like the "least dependency baggage" way of encoding the information

File: server/src/db.rs [hunk 2]
```diff
@@ -1460,7 +1460,7 @@ mod tests {
             .unwrap();
 
         catalog_chunk_size_bytes_metric_eq(&test_db.metric_registry, "moved", 1222).unwrap();
-        catalog_chunk_size_bytes_metric_eq(&test_db.metric_registry, "os", 1897).unwrap(); // now also in OS
+        catalog_chunk_size_bytes_metric_eq(&test_db.metric_registry, "os", 1913).unwrap(); // now also in OS
```
Comment (alamb) [inline server/src/db.rs:1463]: for some reason the size of the summary structure (to which I have added a field)  is also included in the chunk size. Which is fine, if a bit annoying

File: server/src/query_tests/sql.rs [hunk 3]
```diff
@@ -288,20 +288,20 @@ async fn sql_select_from_system_columns() {
...
-        "| 1970-01-01T00 | h2o        | state       | 3     |",
-        "| 1970-01-01T00 | h2o        | temp        | 1     |",
-        "| 1970-01-01T00 | h2o        | time        | 3     |",
-        "| 1970-01-01T00 | o2         | city        | 1     |",
-        "| 1970-01-01T00 | o2         | state       | 2     |",
-        "| 1970-01-01T00 | o2         | temp        | 2     |",
-        "| 1970-01-01T00 | o2         | time        | 2     |",
-        "| 1970-01-01T00 | o2         | reading     | 1     |",
-        "+---------------+------------+-------------+-------+",
+        "+---------------+------------+-------------+-------------+---------------+",
```
Comment (alamb) [inline server/src/query_tests/sql.rs:291]: This shows the end to end effect of this change

File: docs/sql.md [hunk 4]
```diff
@@ -212,3 +214,26 @@ GROUP BY database_name, partition_key, table_name
...
+
+Since IOx uses Apache Arrow's
+[DataFusion](https://github.com/apache/arrow-datafusion) query engine, IOx SQL is mostly the same as DataFusion.
+
+In this section, IOx specific SQL tables, commands, and extensions are documented.
+
+
+## System Tables
+
+In addition to the SQL standard `information_schema`, IOx contains several *system tables* that provide access to IOx specific information. The information in each system table is scoped to that particular database. Cross database queries are not possible due to the design of IOx's security model. Another other process, such as the `observer` mode in the IOx SQL client, must be used for queries on information that spans databases.
```
Comment (carols10cents) [inline docs/sql.md:229]:
```suggestion
In addition to the SQL standard `information_schema`, IOx contains several *system tables* that provide access to IOx specific information. The information in each system table is scoped to that particular database. Cross database queries are not possible due to the design of IOx's security model. Another process, such as the `observer` mode in the IOx SQL client, must be used for queries on information that spans databases.
```

File: docs/sql.md [hunk 5]
```diff
@@ -212,3 +214,26 @@ GROUP BY database_name, partition_key, table_name
...
+
+### `system.chunks`
+`system.chunks` contains information about each IOx storage chunk (which holds part of the data for a table).
+
+TODO: document each column, once they have stabilized.
+
+### `system.columns`
+`system.columns` contains IOx specific schema information about each column in each table, such as which columns was loaded as tags, fields, and timestamps in the InfluxDB data model.
```
Comment (carols10cents) [inline docs/sql.md:237]:
```suggestion
`system.columns` contains IOx specific schema information about each column in each table, such as which columns were loaded as tags, fields, and timestamps in the InfluxDB data model.
```

File: server/src/db/system_tables.rs [hunk 6]
```diff
@@ -142,21 +143,28 @@ fn from_partition_summaries(partitions: Vec<PartitionSummary>) -> Result<RecordB
...
-        ("partition_key", Arc::new(partition_key) as ArrayRef),
-        ("table_name", Arc::new(table_name)),
-        ("column_name", Arc::new(column_name)),
-        ("count", Arc::new(count)),
+    RecordBatch::try_from_iter_with_nullable(vec![
+        ("partition_key", Arc::new(partition_key) as ArrayRef, false),
+        ("table_name", Arc::new(table_name), false),
+        ("column_name", Arc::new(column_name), false),
+        ("column_type", Arc::new(column_type), false),
+        ("influxdb_type", Arc::new(influxdb_type), false),
```
Comment (carols10cents) [inline server/src/db/system_tables.rs:167]: Hm, above has:

```
} else {
   influxdb_type.append_null()?;
}
```

so if I'm understanding the arrow concepts correctly, shouldn't this line have `true` for the third element in the tuple to say that `influxdb_type` *can* be null?

```suggestion
        ("influxdb_type", Arc::new(influxdb_type), true),
```

**Review by carols10cents — COMMENTED** [review]
A few copy edits and a question!

File: server/src/db/system_tables.rs (hunk 6 above)
Comment (alamb) [reply server/src/db/system_tables.rs:167]: I think you are correct. Thank you for the spot. 👍

**Review by domodwyer — APPROVED** [review]
Great idea!

**Pull request merged**